tests/failpoints/cases/test_cmd_epoch_checker.rs

```diff
@@ -0,0 +1,373 @@
+    assert!(!merge_rx.recv().unwrap().get_header().has_error());
+    // The write request fails due to epoch not match.
+    cb_receivers.assert_err();
+
+    // Write request is rejected because the source region is merging.
+    // It's not handled by epoch checker now.
+    let write_req = make_write_req(&mut cluster, b"a");
+    let (cb, cb_receivers) = make_cb(&write_req);
+    cluster
+        .sim
+        .rl()
+        .async_command_on_node(1, write_req, cb)
+        .unwrap();
+    cb_receivers.assert_err();
+
+    // Try to put a key on the target region.
+    let write_req = make_write_req(&mut cluster, b"k");
+    let (cb, cb_receivers) = make_cb(&write_req);
+    cluster
+        .sim
+        .rl()
+        .async_command_on_node(1, write_req, cb)
+        .unwrap();
+    // The write request should be blocked until commit-merge is finished.
+    cb_receivers.assert_not_ready();
+
+    // Wait for region merge done.
+    fail::remove(commit_merge_fp);
+    sleep_ms(100);
+    // The write request fails due to epoch not match.
+    cb_receivers.assert_err();
+
+    // New write request can succeed.
+    let write_req = make_write_req(&mut cluster, b"k");
+    let (cb, cb_receivers) = make_cb(&write_req);
+    cluster
+        .sim
+        .rl()
+        .async_command_on_node(1, write_req, cb)
+        .unwrap();
+    cb_receivers.assert_ok();
+}
+
+#[test]
+fn test_reject_proposal_during_rollback_region_merge() {
+    let mut cluster = new_node_cluster(0, 3);
+    configure_for_merge(&mut cluster);
+    let pd_client = cluster.pd_client.clone();
+    pd_client.disable_default_operator();
+    cluster.run_conf_change();
+
+    let r = cluster.get_region(b"");
+    cluster.must_split(&r, b"k");
+
+    // Don't enter the second phase of region merge.
+    let schedule_merge_fp = "on_schedule_merge";
+    fail::cfg(schedule_merge_fp, "return()").unwrap();
+
+    let source = cluster.get_region(b"");
+    let target = cluster.get_region(b"k");
+    // The call is finished when prepare_merge is applied.
+    cluster.must_try_merge(source.get_id(), target.get_id());
+
+    // Add a peer to trigger rollback.
+    pd_client.must_add_peer(target.get_id(), new_peer(2, 4));
+    cluster.must_put(b"k", b"v");
+    must_get_equal(&cluster.get_engine(1), b"k", b"v");
+
+    // Pause on applying so that rolling back merge is not finished.
+    let rollback_merge_fp = "apply_before_rollback_merge";
+    fail::cfg(rollback_merge_fp, "pause").unwrap();
+    fail::remove(schedule_merge_fp);
+    sleep_ms(100);
+
+    // Write request is rejected because the source region is merging.
+    // It's not handled by epoch checker now.
+    let write_req = make_write_req(&mut cluster, b"a");
+    let (cb, cb_receivers) = make_cb(&write_req);
+    cluster
+        .sim
+        .rl()
+        .async_command_on_node(1, write_req, cb)
+        .unwrap();
+    cb_receivers.assert_err();
+
+    fail::remove(rollback_merge_fp);
+    sleep_ms(100);
```

Comment:
I worry this sleep may lead to an unstable test.
If this time is not enough, the `assert_ok` will fail due to epoch not match.
How about using `must_put`? It will retry until succeed.